common/src/main/java/com/wynntils/core/net/DownloadManager.java

```diff
@@ -0,0 +1,220 @@
+        // no modifications should be done during the lifetime of the game instance
+        registrationLock = true;
+
+        graph = DownloadDependencyGraph.build(registeredDownloads);
+
+        // Dump the graph if the system property is set
+        if (DUMP_GRAPH) {
+            graph.logGraph();
+        }
+    }
+
+    public void download() {
+        // Reset the state of the manager, as the downloads are being started
+        graph.resetState();
+        currentDownloads = new Pair[MAX_PARALLEL_DOWNLOADS];
+
+        // Start the downloads by filling the parallel download slots
+        // After that, the manager will regulate the downloads by itself
+        for (int i = 0; i < MAX_PARALLEL_DOWNLOADS; i++) {
+            QueuedDownload queuedDownload = graph.nextDownload();
+
+            if (queuedDownload == null) {
+                // This may not be an issue, but it can be a sign of a bug, or a bad configuration
+                WynntilsMod.warn("Max parallel downloads was not reached, but there are no more downloads to start.");
+                return;
+            }
+
+            currentDownloads[i] = new Pair<>(queuedDownload, getDownload(queuedDownload));
+        }
+    }
+
+    QueuedDownload queueDownload(UrlId urlId, CoreComponent callerComponent, Dependency dependency) {
+        if (registrationLock) {
+            throw new IllegalStateException("Cannot queue downloads after the download graph is already built.");
+        }
+
+        QueuedDownload queuedDownload = new QueuedDownload(callerComponent, urlId, dependency);
+        registeredDownloads.add(queuedDownload);
+        return queuedDownload;
+    }
+
+    private Download getDownload(QueuedDownload queuedDownload) {
+        Download download = Managers.Net.download(queuedDownload.urlId());
+
+        Consumer<Reader> readerHandler = queuedDownload.onCompletionReader();
+        if (readerHandler != null) {
+            download.handleReader(
+                    wrapDownloadHandler(readerHandler, queuedDownload), wrapDownloadFailure(queuedDownload));
+            return download;
+        }
+
+        Consumer<JsonObject> jsonObjectHandler = queuedDownload.onCompletionJsonObject();
+        if (jsonObjectHandler != null) {
+            download.handleJsonObject(
+                    wrapDownloadHandler(jsonObjectHandler, queuedDownload), wrapDownloadFailure(queuedDownload));
+            return download;
+        }
+
+        Consumer<JsonArray> jsonArrayHandler = queuedDownload.onCompletionJsonArray();
+        if (jsonArrayHandler != null) {
+            download.handleJsonArray(
+                    wrapDownloadHandler(jsonArrayHandler, queuedDownload), wrapDownloadFailure(queuedDownload));
+            return download;
+        }
+
+        throw new IllegalStateException("Queued download has no handler set: " + queuedDownload);
+    }
+
+    private void replaceDownload(QueuedDownload toBeReplaced) {
```

Comment:
The name got me thinking that you first downloaded to a temporary file name, and then moved the file in place one the download was successful. (If we don't do that elsewhere in the NetManager, it is probably something we should do.)

But after scratching my head a bit, I realize this is the code responsible for repopulating the queue. I'd suggest renaming the method. Maybe `queueNextDownload()` instead?